File: app/src/main/java/se/appshack/android/refactoring/PokemonListAdapter.java
package se.appshack.android.refactoring;

import android.app.Activity;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

public class PokemonListAdapter extends RecyclerView.Adapter<PokemonViewHolder> {

    private final Activity activity;
    private List<NamedResponseModel> data;

    public PokemonListAdapter(Activity context, List<NamedResponseModel> pokemonModels) {
        this.activity = context;
        this.data = pokemonModels;
    }

    @NonNull
    @Override
    public PokemonViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new PokemonViewHolder(LayoutInflater.from(activity).inflate(R.layout.viewholder_pokemon_list, null, false));
    }

    @Override
    public void onBindViewHolder(@NonNull PokemonViewHolder pokemonViewHolder, int i) {
        final NamedResponseModel responseModel = data.get(i);

        pokemonViewHolder.bind(responseModel);
        pokemonViewHolder.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent();
                intent.setClass(activity, PokemonDetailsActivity.class);
                intent.putExtra("POKEMON_NAME", responseModel.name);
                intent.putExtra("POKEMON_URL", responseModel.url);
                activity.startActivity(intent);
            }
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }
}

class PokemonViewHolder extends RecyclerView.ViewHolder {

    public PokemonViewHolder(@NonNull View itemView) {
        super(itemView);
    }

    public void bind(NamedResponseModel pokemon) {
        String number = String.format("#%s", pokemon.url.substring(pokemon.url.indexOf("pokemon/") + 8, pokemon.url.length() - 1));
        ((TextView) itemView.findViewById(R.id.pokemon_number)).setText(number);

        String formattedName = pokemon.name.substring(0,1).toUpperCase() + pokemon.name.substring(1);
        ((TextView) itemView.findViewById(R.id.pokemon_name)).setText(formattedName);
    }

    public void setOnClickListener(View.OnClickListener listener) {
        itemView.setOnClickListener(listener);
    }
}

File: app/src/main/java/se/appshack/android/refactoring/main/presentation/ui/adapter/PokemonListAdapter.kt
package se.appshack.android.refactoring.main.presentation.ui.adapter

import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.fragment_pokemon_details.*
import kotlinx.android.synthetic.main.viewholder_pokemon_list.view.*

import se.appshack.android.refactoring.main.data.model.NamedResponseModel
import se.appshack.android.refactoring.R
import se.appshack.android.refactoring.main.data.extensions.getID
import se.appshack.android.refactoring.main.data.model.response.PokemonDetailsResponse
import se.appshack.android.refactoring.main.data.model.response.PokemonSpeciesResponse
import se.appshack.android.refactoring.main.domain.PokemonUseCase

class PokemonListAdapter(
    private val activity: Activity,
    private val data: List<NamedResponseModel>
) : RecyclerView.Adapter<PokemonListAdapter.PokemonViewHolder>() {

    private var onPokemonClicked: ((namedResponseModel: NamedResponseModel) -> Any )? = null

    private val pokemonDetailsCache = HashMap<String, PokemonDetailsResponse>()

    private var pokemonDetailsRequester: ( (pokemonId: String, position: Int) -> Any )? = null

    fun addPokemonDetailsCache(pokemonID: String, pokemonDetails: PokemonDetailsResponse, position: Int){
        pokemonDetailsCache[pokemonID] = pokemonDetails
        notifyItemChanged(position)
    }

    fun setOnPokemonClicked( onPokemonClicked: ((namedResponseModel: NamedResponseModel) -> Any ) ): PokemonListAdapter{
        this.onPokemonClicked = onPokemonClicked
        return this
    }

    fun setPokemonDetailsRequester(pokemonDetailsRequester: ( (pokemonId: String, position: Int) -> Unit )): PokemonListAdapter {
        this.pokemonDetailsRequester = pokemonDetailsRequester
        return this
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): PokemonViewHolder {
        return PokemonViewHolder(LayoutInflater.from(activity).inflate(R.layout.viewholder_pokemon_list, null, false))
    }

    override fun onBindViewHolder(pokemonViewHolder: PokemonViewHolder, i: Int) {
        val responseModel = data[i]

        pokemonViewHolder.bind(responseModel, i)

        pokemonViewHolder.setOnClickListener(View.OnClickListener {
            onPokemonClicked?.invoke(responseModel)
        })
    }

    override fun getItemCount(): Int {
        return data.size
    }

    inner class PokemonViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(pokemon: NamedResponseModel, position: Int) {
            val number = String.format("#%s", pokemon.url?.substring(pokemon.url!!.indexOf("pokemon/") + 8, pokemon.url!!.length - 1))
            (itemView.findViewById<View>(R.id.pokemon_number) as TextView).text = number

            val formattedName = pokemon.name?.substring(0, 1)?.toUpperCase() + pokemon.name?.substring(1)
            (itemView.findViewById<View>(R.id.pokemon_name) as TextView).text = formattedName

            if (pokemonDetailsCache.containsKey(pokemon.getID().toString())){
                itemView.imageFront.visibility = View.VISIBLE
                itemView.animCircle.visibility = View.GONE
                Picasso.get().load(pokemonDetailsCache[pokemon.getID().toString()]?.sprites?.frontDefault).into(itemView.imageFront)
            }else {
                itemView.imageFront.visibility = View.GONE
                itemView.animCircle.visibility = View.VISIBLE
                pokemonDetailsRequester?.invoke(pokemon.getID().toString(), position)
            }

        }

        fun setOnClickListener(listener: View.OnClickListener) {
            itemView.setOnClickListener(listener)
        }
    }

}

